dpnp/dpnp_iface_linearalgebra.py

```diff
@@ -60,6 +61,56 @@
 ]
 
 
+def _gemm_res_dtype(*arrays, casting):
+    """
+    Determines the data types for matmul operation and the output array of matmul operation.
+
+    The output array data type is determined based on the Promotion Type Rule
+    and device capibilities. The data type used in matmul operation is an 'inexact' data type
+    determined based on the output data type and device capabilities.
+    Both data types are determined based on the fact that the output array data type can be cast
+    to the other data type according to casting rule specified, otherwise a ``TypeError`` is raised.
+
+    Parameters
+    ----------
+    arrays : {dpnp_array, usm_ndarray}
+        Input arrays.
+    casting : {'no', 'equiv', 'safe', 'same_kind', 'unsafe'}, optional
+        Controls what kind of data casting may occur.
+
+    Returns
+    -------
+    gemm_dtype, res_dtype :
+        The appropriate data types for performing matmul operation and presenting output array.
+
+    """
+
+    res_dtype = dpnp.result_type(*arrays)
+    gemm_dtype = dpnp.default_float_type(device=arrays[0].device)
+    if dpnp.issubdtype(res_dtype, dpnp.complexfloating):
+        gemm_dtype = (
+            dpnp.complex64 if gemm_dtype == dpnp.float32 else dpnp.complex128
+        )
+
+    if dpnp.can_cast(res_dtype, gemm_dtype, casting):
+        if res_dtype in [
+            dpnp.float64,
+            dpnp.complex128,
+        ]:  # in case device does not support fp64
+            return gemm_dtype, gemm_dtype
+        elif res_dtype in [
```

Comment:
I get the following on Iris Xe gpu which shows device capabilities is not taken into account.
>>> a=dpnp.array([1,2,3], dtype="i4")
>>> b=dpnp.array([1,2,3], dtype="f4")
>>> dpnp.result_type(a,b)
dtype('float64')

Reply:
This is because `dpnp.result_type` wasn't properly updated, should be something like:
```python
def result_type(*arrays_and_dtypes):
    usm_arrays_and_dtypes = [
        dpnp.get_usm_ndarray(X) if isinstance(X, (dpnp_array, dpt.usm_ndarray)) else X
        for X in arrays_and_dtypes
    ]
    return dpt.result_type(*usm_arrays_and_dtypes)
```